### PR TITLE
Correctly resolveReference after we load/create embedded group

### LINKED ISSUE #45872: Join not working as expected with embedded layers/tables
### What is the bug or the crash?

Hi,

I have created a qgis project with two joined layers. I would like to use it as an integrated project. if the layers are contained in a layer group, the child project doesn't expose the joins between the layers.

### Steps to reproduce the issue

To check if the joins can be integrated into a child project

The functional way with no layer group :
1. Open an QGIS project
2. load layers linked by a foreign key
3. join them in the layer's properties
4. save the project
5. Open a new project
6. Integrate all the layers from the previous project
7. Save the new project
8. reload the new project 
9. you will notice the joins are imported
----------------------------------------------
The problematic way with layer group
1. Open an QGIS project
2. load layers related by a foreign key
3. join them
4. Set the joined layers into a layer group
5. save the project
6. Open a new project
7. Integrated all the layers from the previous project
8. Save the new project
9. reload the new project 
10. you will notice the joins are not imported

### Versions

Version de QGIS 3.20.3-Odense
Révision du code 495fbaec
Version de Qt 5.15.2
Version de Python 3.9.5
Version de GDAL/OGR 3.3.1
Version de GEOS 3.9.1-CAPI-1.14.2
Version de PDAL 2.3.0
Version de l'OS Windows 10 Version 2009

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by troopa81**
Fixes #45872 : joins are not working with embedded group/layers
- When loading a project we need to load embedded node before calling resolveReference.
- When adding and embedded group or layer, we need to call resolveReference

### CODE REVIEW

**Comment by lindacamathias** [comment]
hi! should this https://github.com/qgis/QGIS/issues/44162  issue also be solved with your fix?

**Comment by troopa81** [comment]
> hi! should this #44162 issue also be solved with your fix @troopa81 ?

Well, embedded joins seems to work fine as long as they don't belong to a group. So if you have group, the issue #44162 is likely to be fixed with this PR. If not, the better is to test once its merged to see if there is any other issue

**Comment by lindacamathias** [comment]
ok. here are my test results, which may be usefull.
 it did not work with my embedded group. it worked with single embedded layers with joins which where already included in the project. but i was not able to embed new layers into the project (just nothing happened at all). when embedding a new group it wrote just the group name  in the layer panel but no layers where included.

**Comment by troopa81** [comment]
@lindacamathias I confirm that the different issues you describe are all fixed with this PR. We just have to wait for someone to gently review and merge this.

**Comment by lindacamathias** [comment]
hi @troopa81 
i tested your pull version and i can confirm that joins were correctly included in the embedded layer when opening an existing project. that's nice, thanks for that fix! 
When i embed new layers in the project however i still have to do once the copypaste-workaround to get the joins of the layer, and as you mentioned the workaround is still necessary for embedded layers with joins within a group. 

embed_layers

**Comment by m-kuhn** [comment]
@troopa81 do you think the issues described by @lindacamathias are related / solvable in this pull request or are they separate things?

**Comment by troopa81** [comment]
@lindacamathias Are you sure the joins don't exist in your example. If you open the attribute table, do you see joined fields? I just tested it on my side and it works.

Regarding the demo video, I understand you cannot open the property dialog  (hence the joins dialog) from the just embedded layer. That's normal. You cannot change properties of an embedded layer, you have to open the project which define the embedded layer to do so.

When you copy/paste the layer, it is no more and embedded layer (the link icon should disappear though) and so you can see the properties dialog.

@m-kuhn IMHO, there is no extra issue to be fixed

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -13607,52 +13607,61 @@ void QgisApp::addMapLayer( QgsMapLayer *mapLayer )
...
+  // fix broken relations and dependencies
+  for ( const QString &id : constSortedIds )
+  {
+    QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( QgsProject::instance()->mapLayer( id ) );
+    if ( vlayer )
+      vectorLayerStyleLoaded( vlayer, QgsMapLayer::AllStyleCategories );
+  }
+
+  // Resolve references to other layers
+  QMap<QString, QgsMapLayer *> layers = QgsProject::instance()->mapLayers();
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp:13661]:
```suggestion
  const QMap<QString, QgsMapLayer *> layers = QgsProject::instance()->mapLayers();
```

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -13607,52 +13607,61 @@ void QgisApp::addMapLayer( QgsMapLayer *mapLayer )
...
+  for ( QMap<QString, QgsMapLayer *>::iterator it = layers.begin(); it != layers.end(); ++it )
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp:13662]:
```suggestion
  for ( QMap<QString, QgsMapLayer *>::iterator it = layers.const_begin(); it != layers.const_end(); ++it )
```

**Review by m-kuhn — APPROVED** [review]
Looks good to me, thanks !

**Comment by m-kuhn** [comment]
> When you copy/paste the layer, it is no more and embedded layer (the link icon should disappear though) and so you can see the properties dialog.

I have just checked, this symbol is inserted based on a custom property "embedded" of the layer (see `QgsLayerTreeViewEmbeddedIndicatorProvider`). I assume this property is copied with the layer, and that's why the symbol is kept, but that's a different issue.

**Comment by lindacamathias** [comment]
hey @troopa81 i tested again and yes the joins are there in the attribute table. also the ones from embedded layers in a group. there was an error with the joining tables in my testproject, sorry for the circumstances. 
the "embedded symbol" remaining at the layer panel after copy-pasting them did confuse me too.
so the embedded layer issue seem to be well-fixed, thanks a lot !!

File: src/app/qgisapp.cpp [hunk 3]
```diff
@@ -13659,7 +13659,7 @@ void QgisApp::addEmbeddedItems( const QString &projectFile, const QStringList &g
 
   // Resolve references to other layers
   const QMap<QString, QgsMapLayer *> layers = QgsProject::instance()->mapLayers();
-  for ( QMap<QString, QgsMapLayer *>::iterator it = layers.begin(); it != layers.end(); ++it )
+  for ( QMap<QString, QgsMapLayer *>::iterator it = layers.const_begin(); it != layers.const_end(); ++it )
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp:13662]: Sorry, was it `constBegin()`?

Comment (troopa81) [reply src/app/qgisapp.cpp]: Yes :) I fixed it. Thanks for review

**Pull request merged**